meddlr/metrics/lpips_meddlr.py

```diff
@@ -0,0 +1,163 @@
+            raise ModuleNotFoundError(
+                "LPIPS metric requires that lpips is installed."
+                "Either install as `pip install meddlr[metrics]` or `pip install lpips`."
+            )
+
+        valid_net_type = ("vgg", "alex", "squeeze")
+        if net_type not in valid_net_type:
+            raise ValueError(
+                f"Invalid `net_type` ('{net_type}'). Expected one of {valid_net_type}."
+            )
+
+        valid_modes = ("grayscale", "rgb")
+        if mode not in valid_modes:
+            raise ValueError(f"Invalid `mode` ('{mode}'). Expected one of {valid_modes}.")
+
+        self.net = NoTrainLpips(net=net_type, lpips=lpips, verbose=False)
+        self.mode = mode
+
+    def func(self, preds: torch.Tensor, targets: torch.Tensor):
+
+        if self.mode == "grayscale":
+            loss_shape = (targets.shape[0], targets.shape[1])
+
+        elif self.mode == "rgb":
+            if targets.shape[1] != 3:
+                raise ValueError(
+                    f"Channel dimension must have size 3 for rgb mode,\
+                    but got tensor of shape {targets.shape}."
+                )
+
+            is_complex = cplx.is_complex(targets) or cplx.is_complex_as_real(targets)
+            if is_complex:
+                raise TypeError(
+                    f"Data type must be real when mode is {self.mode},\
+                    but got data type {targets.dtype}"
+                )
+
+            loss_shape = (targets.shape[0], 1)
+
+        preds = self.preprocess_lpips(preds)
+        targets = self.preprocess_lpips(targets)
+
+        loss = self.net(preds, targets).squeeze()
+        loss = loss.view(loss_shape)
+
+        return loss
+
+    def preprocess_lpips(self, img: torch.Tensor) -> torch.Tensor:
+        """
+        Preprocess image per LPIPS implementation.
+
+        Converts images to magnitude images if complex and normalizes between [-1, 1].
+        If self.mode is 'grayscale', then each channel dimension will be replicated 3 times.
+
+        Args:
+            img (torch.Tensor): Tensor to preprocess.
+
+        Returns:
+            img (torch.Tensor): Preprocessed tensor.
+        """
+
+        is_complex = cplx.is_complex(img) or cplx.is_complex_as_real(img)
+        if is_complex:
+            img = cplx.abs(img)
+
+        if self.mode == "grayscale":
+            shape = (img.shape[0], img.shape[1], -1)  # normalize over batch and channel dimensions
+            img_min = torch.amin(img.reshape(shape), dim=-1, keepdim=True).unsqueeze(-1)
+            img_max = torch.amax(img.reshape(shape), dim=-1, keepdim=True).unsqueeze(-1)
+            img = 2 * (img - img_min) / (img_max - img_min) - 1
+
+            img = img.reshape(img.shape[0] * img.shape[1], 1, img.shape[2], img.shape[3])
+            img = img.repeat(1, 3, 1, 1)
+
+        elif self.mode == "rgb":
+            shape = (img.shape[0], -1)  # normalize over batch dimension only
```

Comment:
this isn't normalizing over the batch - each example is being normalized separately